File: runtime/android/NexaDevSchema.kt
package __NEXA_PACKAGE__

/**
 * Canonical schema definitions, protocol version constants, and decoder keys
 * shared between Nexa development servers and Android dev runtimes.
 */
internal object NexaDevSchema {
    const val PROTOCOL_VERSION: Int = 3
    const val DEV_IR_FORMAT_VERSION: Int = 1
    const val TARGET_PLATFORM: String = "android"
}

/**
 * Strongly-typed string constants for dynamic Dev IR JSON keys and message types.
 */
internal object NexaDevKeys {
    // Protocol envelope keys
    const val TYPE = "type"
    const val PAYLOAD = "payload"
    const val PROTOCOL_VERSION = "protocol_version"
    const val SESSION_TOKEN = "session_token"
    const val TARGET = "target"
    const val REVISION = "revision"
    const val BASE_REVISION = "base_revision"
    const val MODULE = "module"
    const val OPERATIONS = "operations"
    const val IDENTITIES = "identities"
    const val DIAGNOSTICS = "diagnostics"
    const val ENABLED = "enabled"
    const val REASON = "reason"

    // Client message types
    const val MSG_HELLO = "hello"
    const val MSG_ACKNOWLEDGE = "acknowledge"
    const val MSG_REQUEST_FULL_MODULE = "request_full_module"
    const val MSG_DISCONNECT = "disconnect"

    // Server message types
    const val MSG_WELCOME = "welcome"
    const val MSG_FULL_MODULE = "full_module"
    const val MSG_PATCH = "patch"
    const val MSG_DIAGNOSTICS = "diagnostics"
    const val MSG_RELOAD = "reload"
    const val MSG_RESTART = "restart"
    const val MSG_PERFORMANCE_OVERLAY = "performance_overlay"

    // Patch operations
    const val OP = "op"
    const val OP_SET = "set"
    const val OP_REMOVE = "remove"
    const val OP_PATH = "path"
    const val OP_VALUE = "value"

    // Identity fields
    const val ID = "id"
    const val KIND = "kind"

    // Module fields
    const val APP_NAME = "app_name"
    const val BODY = "body"
    const val SCREENS = "screens"
    const val STATES = "states"
    const val FUNCTIONS = "functions"
    const val STRUCTS = "structs"
    const val COMPONENTS = "components"
    const val DIRECTION = "direction"
    const val STATUS_BAR = "status_bar"
    const val ON_APPEAR = "on_appear"
    const val ON_APPEAR_ASYNC = "on_appear_async"
    const val ON_DISAPPEAR = "on_disappear"
    const val ON_ACTIVE = "on_active"
    const val ON_INACTIVE = "on_inactive"
    const val ON_BACKGROUND = "on_background"

    // Screen fields
    const val NAME = "name"
    const val PARAMETERS = "parameters"

    // State declaration fields
    const val TY = "ty"
    const val INITIAL = "initial"

    // Diagnostic fields
    const val FILE = "file"
    const val LINE = "line"
    const val COLUMN = "column"
    const val MESSAGE = "message"
    const val SEVERITY = "severity"

    // Native API namespaces
    const val NS_NETWORK = "Network"
    const val NS_FILE = "File"
    const val NS_PATH = "Path"
    const val NS_PERMISSIONS = "Permissions"
}

internal object NexaDevNodeKind {
    const val LAYOUT = "Layout"
    const val COLUMN = "Column"
    const val ROW = "Row"
    const val STACK = "Stack"
    const val TEXT = "Text"
    const val BUTTON = "Button"
    const val PRESSABLE = "Pressable"
    const val TEXT_INPUT = "TextInput"
    const val FAST_LIST = "FastList"
    const val SWITCH = "Switch"
    const val IMAGE = "Image"
    const val REFRESH_CONTROL = "RefreshControl"
    const val IF = "If"
    const val WHEN = "When"
    const val LINK = "Link"
    const val ACCESSIBILITY = "Accessibility"
    const val KEYBOARD_AWARE = "KeyboardAware"
    const val APP_BOTTOM_BAR = "AppBottomBar"
    const val BOTTOM_SHEET = "BottomSheet"
    const val NAVIGATION_STACK = "NavigationStack"
    const val NAVIGATION_LINK = "NavigationLink"
    const val NAVIGATION_BACK = "NavigationBack"
    const val CONTENT = "Content"
    const val COMPONENT_CALL = "ComponentCall"
}

internal object NexaDevActionKind {
    const val ASSIGN = "Assign"
    const val EXPRESSION = "Expression"
    const val IF = "If"
    const val FOR = "For"
    const val FOR_MAP = "ForMap"
    const val WHILE = "While"
    const val COLLECTION_MUTATION = "CollectionMutation"
    const val TRY_CATCH = "TryCatch"
    const val BREAK = "Break"
    const val CONTINUE = "Continue"
    const val NATIVE_CALL = "NativeCall"
}

internal object NexaDevCollectionMutationKind {
    const val ARRAY_APPEND = "ArrayAppend"
    const val ARRAY_REMOVE_AT = "ArrayRemoveAt"
    const val SET_INSERT = "SetInsert"
    const val SET_REMOVE = "SetRemove"
    const val MAP_SET = "MapSet"
    const val MAP_REMOVE = "MapRemove"
}

internal object NexaDevExprKind {
    const val STRING = "String"
    const val BOOL = "Bool"
    const val NUMBER = "Number"
    const val ARRAY = "Array"
    const val SET = "Set"
    const val MAP = "Map"
    const val PAIR = "Pair"
    const val TRIPLE = "Triple"
    const val ENUM_VALUE = "EnumValue"
    const val STATE = "State"
    const val INTERPOLATION = "Interpolation"
    const val ADD = "Add"
    const val NOT = "Not"
    const val BINARY = "Binary"
    const val CONTAINS = "Contains"
    const val COLLECTION_TRANSFORM = "CollectionTransform"
    const val CLOSURE = "Closure"
    const val INDEX = "Index"
    const val MEMBER = "Member"
    const val RANGE = "Range"
    const val NULL = "Null"
    const val COALESCE = "Coalesce"
    const val RESULT_OK = "ResultOk"
    const val RESULT_ERR = "ResultErr"
    const val TRY = "Try"
    const val CALL = "Call"
    const val NATIVE_CALL = "NativeCall"
    const val NETWORK_FETCH = "NetworkFetch"
    const val NETWORK_DOWNLOAD = "NetworkDownload"
    const val PATH_JOIN = "PathJoin"
    const val FILE_EXISTS = "FileExists"
    const val FILE_READ_TEXT = "FileReadText"
    const val FILE_WRITE_TEXT = "FileWriteText"
    const val FILE_DELETE = "FileDelete"
    const val PERMISSION_OP = "PermissionOp"
    const val AWAIT = "Await"
    const val TRY_AWAIT = "TryAwait"
}

File: runtime/ios/NexaDevSchema.swift
import Foundation

/// Canonical schema definitions, protocol version constants, and decoder keys
/// shared between Nexa development servers and native dev runtimes.
public enum NexaDevSchema {
    /// Dev protocol version exchanged during WebSocket handshake.
    public static let protocolVersion: Int = 3

    /// Format version of the Dev IR payload.
    public static let devIRFormatVersion: Int = 1

    /// Platform identifier for iOS dev runtimes.
    public static let targetPlatform = "ios"
}

/// Strongly-typed string constants for dynamic Dev IR dictionary keys and message types.
public enum NexaDevKeys {
    // Protocol envelope keys
    public static let type = "type"
    public static let payload = "payload"
    public static let protocolVersion = "protocol_version"
    public static let sessionToken = "session_token"
    public static let target = "target"
    public static let revision = "revision"
    public static let baseRevision = "base_revision"
    public static let module = "module"
    public static let operations = "operations"
    public static let identities = "identities"
    public static let diagnostics = "diagnostics"
    public static let enabled = "enabled"
    public static let reason = "reason"

    // Client message types
    public static let msgHello = "hello"
    public static let msgAcknowledge = "acknowledge"
    public static let msgRequestFullModule = "request_full_module"
    public static let msgDisconnect = "disconnect"

    // Server message types
    public static let msgWelcome = "welcome"
    public static let msgFullModule = "full_module"
    public static let msgPatch = "patch"
    public static let msgDiagnostics = "diagnostics"
    public static let msgReload = "reload"
    public static let msgRestart = "restart"
    public static let msgPerformanceOverlay = "performance_overlay"

    // Patch operations
    public static let op = "op"
    public static let opSet = "set"
    public static let opRemove = "remove"
    public static let opPath = "path"
    public static let opValue = "value"

    // Identity fields
    public static let id = "id"
    public static let kind = "kind"

    // Module fields
    public static let appName = "app_name"
    public static let body = "body"
    public static let screens = "screens"
    public static let states = "states"
    public static let functions = "functions"
    public static let structs = "structs"
    public static let components = "components"
    public static let direction = "direction"
    public static let statusBar = "status_bar"
    public static let onAppear = "on_appear"
    public static let onAppearAsync = "on_appear_async"
    public static let onDisappear = "on_disappear"
    public static let onActive = "on_active"
    public static let onInactive = "on_inactive"
    public static let onBackground = "on_background"

    // Screen fields
    public static let name = "name"
    public static let parameters = "parameters"

    // State declaration fields
    public static let ty = "ty"
    public static let initial = "initial"

    // Diagnostic fields
    public static let file = "file"
    public static let line = "line"
    public static let column = "column"
    public static let message = "message"
    public static let severity = "severity"

    // Native API namespaces
    public static let nsNetwork = "Network"
    public static let nsFile = "File"
    public static let nsPath = "Path"
    public static let nsPermissions = "Permissions"
}

/// Known node kinds in Nexa Dev IR.
public enum NexaDevNodeKind {
    public static let layout = "Layout"
    public static let column = "Column"
    public static let row = "Row"
    public static let stack = "Stack"
    public static let text = "Text"
    public static let button = "Button"
    public static let pressable = "Pressable"
    public static let textInput = "TextInput"
    public static let fastList = "FastList"
    public static let switchNode = "Switch"
    public static let image = "Image"
    public static let refreshControl = "RefreshControl"
    public static let ifNode = "If"
    public static let whenNode = "When"
    public static let link = "Link"
    public static let accessibility = "Accessibility"
    public static let keyboardAware = "KeyboardAware"
    public static let appBottomBar = "AppBottomBar"
    public static let bottomSheet = "BottomSheet"
    public static let navigationStack = "NavigationStack"
    public static let navigationLink = "NavigationLink"
    public static let navigationBack = "NavigationBack"
    public static let content = "Content"
    public static let componentCall = "ComponentCall"
}

/// Known action kinds in Nexa Dev IR.
public enum NexaDevActionKind {
    public static let assign = "Assign"
    public static let expression = "Expression"
    public static let ifAction = "If"
    public static let forLoop = "For"
    public static let forMap = "ForMap"
    public static let whileLoop = "While"
    public static let collectionMutation = "CollectionMutation"
    public static let tryCatch = "TryCatch"
    public static let breakAction = "Break"
    public static let continueAction = "Continue"
    public static let nativeCall = "NativeCall"
}

/// Known collection mutation kinds in Nexa Dev IR.
public enum NexaDevCollectionMutationKind {
    public static let arrayAppend = "ArrayAppend"
    public static let arrayRemoveAt = "ArrayRemoveAt"
    public static let setInsert = "SetInsert"
    public static let setRemove = "SetRemove"
    public static let mapSet = "MapSet"
    public static let mapRemove = "MapRemove"
}

/// Known expression kinds in Nexa Dev IR.
public enum NexaDevExprKind {
    public static let literalString = "String"
    public static let literalBool = "Bool"
    public static let literalNumber = "Number"
    public static let array = "Array"
    public static let set = "Set"
    public static let map = "Map"
    public static let pair = "Pair"
    public static let triple = "Triple"
    public static let enumValue = "EnumValue"
    public static let state = "State"
    public static let interpolation = "Interpolation"
    public static let add = "Add"
    public static let not = "Not"
    public static let binary = "Binary"
    public static let contains = "Contains"
    public static let collectionTransform = "CollectionTransform"
    public static let closure = "Closure"
    public static let index = "Index"
    public static let member = "Member"
    public static let range = "Range"
    public static let null = "Null"
    public static let coalesce = "Coalesce"
    public static let resultOk = "ResultOk"
    public static let resultErr = "ResultErr"
    public static let tryExpr = "Try"
    public static let call = "Call"
    public static let nativeCall = "NativeCall"
    public static let networkFetch = "NetworkFetch"
    public static let networkDownload = "NetworkDownload"
    public static let pathJoin = "PathJoin"
    public static let fileExists = "FileExists"
    public static let fileReadText = "FileReadText"
    public static let fileWriteText = "FileWriteText"
    public static let fileDelete = "FileDelete"
    public static let permissionOp = "PermissionOp"
    public static let awaitExpr = "Await"
    public static let tryAwait = "TryAwait"
}
